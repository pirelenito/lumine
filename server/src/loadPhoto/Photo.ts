export default interface Photo {
  relativePath: string
  contentHash: string

  // preview: {
  //   thumbnailPath: string
  //   fullsizePath: string
  // }

  // metadata: {
  //   file: {
  //     ctimeMs: number
  //     mtimeMs: number
  //     size: number
  //   }
  // }
}
