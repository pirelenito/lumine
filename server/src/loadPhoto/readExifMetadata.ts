import { pickBy, type, contains } from 'ramda'
import { ExifImage } from 'exif'
import { ExifMetadata } from './Photo'

const onlyNumbersAndStrings = pickBy((val, key) => contains(type(val), ['Number', 'String']))

export default (fullPath: string): Promise<ExifMetadata> =>
  new Promise((resolve, reject) => {
    try {
      new ExifImage({ image: fullPath }, (err: any, metadata = { image: {}, exif: {}, gps: {} }) => {
        if (err && err.message !== 'No Exif segment found in the given image.') {
          return reject(err)
        }

        const rawMetadata = {
          image: onlyNumbersAndStrings(metadata.image),
          exif: onlyNumbersAndStrings(metadata.exif),
          gps: onlyNumbersAndStrings(metadata.gps),
        }

        resolve(rawMetadata)
      })
    } catch (e) {
      reject(e)
    }
  })
