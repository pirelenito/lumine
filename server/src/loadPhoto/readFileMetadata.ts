import { FileMetadata } from './Photo'

const promisify = require('util').promisify
const stat = promisify(require('fs').stat)

export default async (fullPath: string): Promise<FileMetadata> => {
  const fileStat = await stat(fullPath)

  return {
    // birthtime can be null on some filesystems
    // which makes node return something else that might be wrong
    createdAt: convertToDate(fileStat.birthtimeMs > fileStat.mtimeMs ? fileStat.mtimeMs : fileStat.birthtimeMs),
    modifiedAt: convertToDate(fileStat.mtimeMs),
    size: fileStat.size,
  }
}

const convertToDate = (timeMs: number) => new Date(Math.floor(timeMs / 1000) * 1000)
