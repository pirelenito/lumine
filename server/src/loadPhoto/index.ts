import Photo from './Photo'
import { join } from 'path'
import generatePreview from './generatePreview'
import calculateContentHash from './calculateContentHash'
import readFileMetadata from './readFileMetadata'

interface Config {
  libraryBasePath: string
  cacheBasePath: string
}

export default (config: Config) => async (relativePath: string): Promise<Photo> => {
  const fullPath = join(config.libraryBasePath, relativePath)
  const contentHash = await calculateContentHash(fullPath)
  const fileMetadata = await readFileMetadata(fullPath)
  const preview = await generatePreview(config.cacheBasePath)(contentHash, fullPath)

  return { relativePath, contentHash, metadata: { file: fileMetadata }, preview }
}
