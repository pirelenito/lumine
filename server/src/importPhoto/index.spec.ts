import { join } from 'path'
import { readdirSync } from 'fs'
import rimraf from 'rimraf'

import loadPhoto from '.'

const libraryBasePath = join(__dirname, '../../fixtures')
const cacheBasePath = join(__dirname, '../../tmp')
const config = { libraryBasePath, cacheBasePath, httpPort: 80 }

beforeAll(() => rimraf.sync(cacheBasePath))

readdirSync(libraryBasePath)
  .filter(photoPath => photoPath.match(/\.(arw|jpg|jpeg|)$/i))
  .forEach((photoPath: string) => {
    it(`loads ${photoPath}`, async () => {
      const photo = await loadPhoto(config)(photoPath)
      expect(photo).toMatchSnapshot()
    }, 180000)
  })
