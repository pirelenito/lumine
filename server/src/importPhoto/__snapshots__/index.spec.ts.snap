// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads canonS2is.jpg 1`] = `
Object {
  "contentHash": "7d5bd5ffdad4b4c71e455631ebc137f189d50ed4",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1187099113000,
      },
      "exif": Object {
        "ApertureValue": 6,
        "ColorSpace": 1,
        "CompressedBitsPerPixel": 3,
        "CreateDate": "2007:08:14 13:45:13",
        "CustomRendered": 0,
        "DateTimeOriginal": "2007:08:14 13:45:13",
        "DigitalZoomRatio": 1,
        "ExifImageHeight": 1944,
        "ExifImageWidth": 2592,
        "ExposureCompensation": 0,
        "ExposureMode": 0,
        "ExposureTime": 0.002,
        "FNumber": 8,
        "Flash": 73,
        "FocalLength": 72,
        "FocalPlaneResolutionUnit": 2,
        "FocalPlaneXResolution": 11520,
        "FocalPlaneYResolution": 11571.42857142857,
        "InteropOffset": 2174,
        "MaxApertureValue": 3.625,
        "MeteringMode": 2,
        "SceneCaptureType": 0,
        "SensingMethod": 2,
        "ShutterSpeedValue": 8.96875,
        "WhiteBalance": 1,
      },
      "gps": Object {},
      "image": Object {
        "ExifOffset": 196,
        "Make": "Canon",
        "Model": "Canon PowerShot S2 IS",
        "ModifyDate": "2007:08:14 13:45:13",
        "Orientation": 1,
        "ResolutionUnit": 2,
        "XResolution": 180,
        "YCbCrPositioning": 1,
        "YResolution": 180,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 1287333,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/7d/5bd5ffdad4b4c71e455631ebc137f189d50ed4.jpg",
    "thumbnailPath": "thumbnail/7d/5bd5ffdad4b4c71e455631ebc137f189d50ed4.jpg",
  },
  "relativePath": "canonS2is.jpg",
}
`;

exports[`loads iphone3gs.JPG 1`] = `
Object {
  "contentHash": "9bb08d3051b22a416aa8eded126a800e9f6d2dd7",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1320507188000,
      },
      "exif": Object {
        "ApertureValue": 2.970853573907009,
        "BrightnessValue": 9.452459016393442,
        "ColorSpace": 1,
        "CreateDate": "2011:11:05 15:33:08",
        "DateTimeOriginal": "2011:11:05 15:33:08",
        "ExifImageHeight": 1536,
        "ExifImageWidth": 2048,
        "ExposureMode": 0,
        "ExposureProgram": 2,
        "ExposureTime": 0.0005521811154058532,
        "FNumber": 2.8,
        "Flash": 32,
        "FocalLength": 3.85,
        "ISO": 64,
        "MeteringMode": 1,
        "SceneCaptureType": 0,
        "SensingMethod": 2,
        "Sharpness": 1,
        "ShutterSpeedValue": 10.822407628128724,
        "WhiteBalance": 0,
      },
      "gps": Object {
        "GPSAltitude": 33.558791801510246,
        "GPSAltitudeRef": 0,
        "GPSLatitudeRef": "S",
        "GPSLongitudeRef": "W",
      },
      "image": Object {
        "ExifOffset": 200,
        "GPSInfo": 590,
        "Make": "Apple",
        "Model": "iPhone 3GS",
        "ModifyDate": "2011:11:05 15:33:08",
        "Orientation": 3,
        "ResolutionUnit": 2,
        "Software": "5.0",
        "XResolution": 72,
        "YCbCrPositioning": 1,
        "YResolution": 72,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 1214856,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/9b/b08d3051b22a416aa8eded126a800e9f6d2dd7.jpg",
    "thumbnailPath": "thumbnail/9b/b08d3051b22a416aa8eded126a800e9f6d2dd7.jpg",
  },
  "relativePath": "iphone3gs.JPG",
}
`;

exports[`loads iphone5s.jpg 1`] = `
Object {
  "contentHash": "e2fe6fcbf4044046933845a233b1129bfd3ec4d8",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1391601952000,
      },
      "exif": Object {
        "ApertureValue": 2.2750072066878064,
        "BrightnessValue": 8.475824175824176,
        "ColorSpace": 1,
        "CreateDate": "2014:02:05 12:05:52",
        "DateTimeOriginal": "2014:02:05 12:05:52",
        "ExifImageHeight": 2448,
        "ExifImageWidth": 3264,
        "ExposureMode": 0,
        "ExposureProgram": 2,
        "ExposureTime": 0.001145475372279496,
        "FNumber": 2.2,
        "Flash": 24,
        "FocalLength": 4.12,
        "FocalLengthIn35mmFormat": 30,
        "ISO": 32,
        "LensMake": "Apple",
        "LensModel": "iPhone 5s back camera 4.12mm f/2.2",
        "MeteringMode": 5,
        "SceneCaptureType": 0,
        "SensingMethod": 2,
        "ShutterSpeedValue": 9.77043673012318,
        "SubSecTimeDigitized": "395",
        "SubSecTimeOriginal": "395",
        "WhiteBalance": 0,
      },
      "gps": Object {
        "GPSAltitude": 61.07607607607608,
        "GPSAltitudeRef": 0,
        "GPSDateStamp": "2014:02:05",
        "GPSImgDirection": 222.1251862891207,
        "GPSImgDirectionRef": "T",
        "GPSLatitudeRef": "N",
        "GPSLongitudeRef": "E",
      },
      "image": Object {
        "ExifOffset": 192,
        "GPSInfo": 1506,
        "Make": "Apple",
        "Model": "iPhone 5s",
        "ModifyDate": "2014:02:05 12:05:52",
        "ResolutionUnit": 2,
        "Software": "7.0.4",
        "XResolution": 72,
        "YCbCrPositioning": 1,
        "YResolution": 72,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 1313425,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/e2/fe6fcbf4044046933845a233b1129bfd3ec4d8.jpg",
    "thumbnailPath": "thumbnail/e2/fe6fcbf4044046933845a233b1129bfd3ec4d8.jpg",
  },
  "relativePath": "iphone5s.jpg",
}
`;

exports[`loads iphone6.jpg 1`] = `
Object {
  "contentHash": "e56e56185e42ac9569b4d811adb5cd495fbcebec",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1533837990000,
      },
      "exif": Object {
        "ApertureValue": 2.2750071245369052,
        "BrightnessValue": 11.340301003344482,
        "ColorSpace": 1,
        "CreateDate": "2018:08:09 18:06:30",
        "DateTimeOriginal": "2018:08:09 18:06:30",
        "ExifImageHeight": 1668,
        "ExifImageWidth": 2968,
        "ExposureCompensation": 0,
        "ExposureMode": 0,
        "ExposureProgram": 2,
        "ExposureTime": 0.00032797638570022957,
        "FNumber": 2.2,
        "Flash": 24,
        "FocalLength": 4.15,
        "FocalLengthIn35mmFormat": 36,
        "ISO": 32,
        "LensMake": "Apple",
        "LensModel": "iPhone 6 back camera 4.15mm f/2.2",
        "MeteringMode": 5,
        "SceneCaptureType": 0,
        "SensingMethod": 2,
        "ShutterSpeedValue": 11.574016563146998,
        "SubSecTimeDigitized": "093",
        "SubSecTimeOriginal": "093",
        "WhiteBalance": 0,
      },
      "gps": Object {},
      "image": Object {
        "ExifOffset": 194,
        "Make": "Apple",
        "Model": "iPhone 6",
        "ModifyDate": "2018:08:09 18:06:30",
        "Orientation": 1,
        "ResolutionUnit": 2,
        "Software": "11.4.1",
        "XResolution": 72,
        "YCbCrPositioning": 1,
        "YResolution": 72,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 993124,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/e5/6e56185e42ac9569b4d811adb5cd495fbcebec.jpg",
    "thumbnailPath": "thumbnail/e5/6e56185e42ac9569b4d811adb5cd495fbcebec.jpg",
  },
  "relativePath": "iphone6.jpg",
}
`;

exports[`loads nokia5.jpg 1`] = `
Object {
  "contentHash": "da1fb28e2764fcd4edc27c4c65167a1349c9b802",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1541680918000,
      },
      "exif": Object {
        "ApertureValue": 2,
        "BrightnessValue": 0.3,
        "ColorSpace": 1,
        "CreateDate": "2018:11:08 12:41:58",
        "DateTimeOriginal": "2018:11:08 12:41:58",
        "ExifImageHeight": 3120,
        "ExifImageWidth": 4160,
        "ExposureMode": 0,
        "ExposureProgram": 0,
        "ExposureTime": 0.030303030303030304,
        "FNumber": 2,
        "Flash": 16,
        "FocalLength": 3.65,
        "FocalLengthIn35mmFormat": 0,
        "ISO": 320,
        "InteropOffset": 759,
        "MeteringMode": 1,
        "SceneCaptureType": 0,
        "SensingMethod": 2,
        "ShutterSpeedValue": 5.059,
        "SubSecTime": "931367",
        "SubSecTimeDigitized": "931367",
        "SubSecTimeOriginal": "931367",
        "WhiteBalance": 0,
      },
      "gps": Object {
        "GPSAltitudeRef": 2,
        "GPSDateStamp": "2018:11:08",
      },
      "image": Object {
        "ExifOffset": 242,
        "GPSInfo": 789,
        "Make": "",
        "Model": "Nokia 5",
        "ModifyDate": "2018:11:08 12:41:58",
        "ResolutionUnit": 2,
        "Software": "TA-1053_00WW-user 8.1.0 OPR1.170623.026 00WW_5_63D release-keys",
        "XResolution": 72,
        "YCbCrPositioning": 1,
        "YResolution": 72,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 2273365,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/da/1fb28e2764fcd4edc27c4c65167a1349c9b802.jpg",
    "thumbnailPath": "thumbnail/da/1fb28e2764fcd4edc27c4c65167a1349c9b802.jpg",
  },
  "relativePath": "nokia5.jpg",
}
`;

exports[`loads nokia620.jpg 1`] = `
Object {
  "contentHash": "a4461089301c564665b392ba4190c6e19458966a",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1365181513000,
      },
      "exif": Object {
        "ColorSpace": 1,
        "CreateDate": "2013:04:05 17:05:13",
        "DateTimeOriginal": "2013:04:05 17:05:13",
        "ExifImageHeight": 1456,
        "ExifImageWidth": 2592,
        "ExposureCompensation": 0,
        "ExposureTime": 0.001095,
        "FNumber": 2.4,
        "Flash": 16,
        "ISO": 100,
        "InteropOffset": 9816,
      },
      "gps": Object {
        "GPSAltitude": 0.007704160246533128,
        "GPSAltitudeRef": 0,
        "GPSDOP": 0.25,
        "GPSLatitudeRef": "S",
        "GPSLongitudeRef": "W",
        "GPSMeasureMode": "3",
      },
      "image": Object {
        "ExifOffset": 2252,
        "GPSInfo": 9846,
        "Make": "Nokia",
        "Model": "Lumia 620",
        "Orientation": 1,
        "ResolutionUnit": 2,
        "Software": "Windows Phone",
        "XResolution": 72,
        "YCbCrPositioning": 1,
        "YResolution": 72,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 1437789,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/a4/461089301c564665b392ba4190c6e19458966a.jpg",
    "thumbnailPath": "thumbnail/a4/461089301c564665b392ba4190c6e19458966a.jpg",
  },
  "relativePath": "nokia620.jpg",
}
`;

exports[`loads nokia5130.jpg 1`] = `
Object {
  "contentHash": "0a434b8b8aa7b45d9b15c809e98e2d4f8f7077ba",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1256217484000,
      },
      "exif": Object {
        "ColorSpace": 1,
        "CreateDate": "2009:10:22 13:18:04",
        "CustomRendered": 0,
        "DateTimeOriginal": "2009:10:22 13:18:04",
        "DigitalZoomRatio": 1,
        "ExifImageHeight": 1600,
        "ExifImageWidth": 1200,
        "ExposureMode": 0,
        "SceneCaptureType": 0,
        "WhiteBalance": 0,
      },
      "gps": Object {},
      "image": Object {
        "ExifOffset": 160,
        "Make": "Nokia",
        "Model": "5130c-2",
        "Orientation": 1,
        "ResolutionUnit": 2,
        "Software": "V 06.65",
        "XResolution": 300,
        "YCbCrPositioning": 1,
        "YResolution": 300,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 295880,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/0a/434b8b8aa7b45d9b15c809e98e2d4f8f7077ba.jpg",
    "thumbnailPath": "thumbnail/0a/434b8b8aa7b45d9b15c809e98e2d4f8f7077ba.jpg",
  },
  "relativePath": "nokia5130.jpg",
}
`;

exports[`loads sonyA6000.arw 1`] = `
Object {
  "contentHash": "98ce370e686f5a40a02523058679a06ca3ff4ce2",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1534255754000,
      },
      "exif": Object {
        "BrightnessValue": 8.946875,
        "ColorSpace": 1,
        "CompressedBitsPerPixel": 8,
        "Contrast": 0,
        "CreateDate": "2018:08:14 14:09:14",
        "CustomRendered": 0,
        "DateTimeOriginal": "2018:08:14 14:09:14",
        "DigitalZoomRatio": 1,
        "ExifImageHeight": 4000,
        "ExifImageWidth": 6000,
        "ExposureCompensation": 0,
        "ExposureMode": 0,
        "ExposureProgram": 3,
        "ExposureTime": 0.00125,
        "FNumber": 3.2,
        "Flash": 16,
        "FocalLength": 35,
        "FocalLengthIn35mmFormat": 52,
        "ISO": 100,
        "InteropOffset": 38114,
        "LensModel": "E 35mm F1.8 OSS",
        "LightSource": 0,
        "MaxApertureValue": 1.6953125,
        "MeteringMode": 5,
        "RecommendedExposureIndex": 100,
        "Saturation": 0,
        "SceneCaptureType": 0,
        "SensitivityType": 2,
        "Sharpness": 0,
        "WhiteBalance": 0,
      },
      "gps": Object {},
      "image": Object {
        "ExifOffset": 348,
        "ImageDescription": "                               ",
        "Make": "SONY",
        "Model": "ILCE-6000",
        "ModifyDate": "2018:08:14 14:09:14",
        "Orientation": 1,
        "ProcessingSoftware": "UFRaw 0.20",
        "Software": "ILCE-6000 v3.20",
        "SubfileType": 1,
        "YCbCrPositioning": 2,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 25133056,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/98/ce370e686f5a40a02523058679a06ca3ff4ce2.jpg",
    "thumbnailPath": "thumbnail/98/ce370e686f5a40a02523058679a06ca3ff4ce2.jpg",
  },
  "relativePath": "sonyA6000.arw",
}
`;

exports[`loads sonyNex3.arw 1`] = `
Object {
  "contentHash": "c78d2b525bb9bfe21f036c21202dcae2fdde3fcb",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1391623582000,
      },
      "exif": Object {
        "BrightnessValue": -0.4203125,
        "ColorSpace": 1,
        "CompressedBitsPerPixel": 8,
        "Contrast": 0,
        "CreateDate": "2014:02:05 18:06:22",
        "CustomRendered": 0,
        "DateTimeOriginal": "2014:02:05 18:06:22",
        "DigitalZoomRatio": 1,
        "ExifImageHeight": 3264,
        "ExifImageWidth": 4912,
        "ExposureCompensation": 0,
        "ExposureMode": 0,
        "ExposureProgram": 3,
        "ExposureTime": 0.02,
        "FNumber": 4.5,
        "Flash": 16,
        "FocalLength": 26,
        "FocalLengthIn35mmFormat": 39,
        "ISO": 3200,
        "InteropOffset": 36704,
        "LensModel": "E PZ 16-50mm F3.5-5.6 OSS",
        "LightSource": 10,
        "MaxApertureValue": 4.3359375,
        "MeteringMode": 5,
        "RecommendedExposureIndex": 3200,
        "Saturation": 0,
        "SceneCaptureType": 0,
        "SensitivityType": 2,
        "Sharpness": 0,
        "WhiteBalance": 1,
      },
      "gps": Object {},
      "image": Object {
        "ExifOffset": 344,
        "ImageDescription": "                               ",
        "Make": "SONY",
        "Model": "NEX-3N",
        "ModifyDate": "2014:02:05 18:06:22",
        "Orientation": 1,
        "ProcessingSoftware": "UFRaw 0.20",
        "Software": "NEX-3N v1.00",
        "SubfileType": 1,
        "YCbCrPositioning": 2,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 17006592,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/c7/8d2b525bb9bfe21f036c21202dcae2fdde3fcb.jpg",
    "thumbnailPath": "thumbnail/c7/8d2b525bb9bfe21f036c21202dcae2fdde3fcb.jpg",
  },
  "relativePath": "sonyNex3.arw",
}
`;

exports[`loads sonyNex3.jpg 1`] = `
Object {
  "contentHash": "6c4e0614f36b68ddf1ef24b938691e48723f71dc",
  "metadata": Object {
    "exif": Object {
      "dates": Object {
        "createdAt": 1391623582000,
      },
      "exif": Object {
        "BrightnessValue": -0.4203125,
        "ColorSpace": 1,
        "CompressedBitsPerPixel": 2,
        "Contrast": 0,
        "CreateDate": "2014:02:05 18:06:22",
        "CustomRendered": 0,
        "DateTimeOriginal": "2014:02:05 18:06:22",
        "DigitalZoomRatio": 1,
        "ExifImageHeight": 3264,
        "ExifImageWidth": 4912,
        "ExposureCompensation": 0,
        "ExposureMode": 0,
        "ExposureProgram": 3,
        "ExposureTime": 0.02,
        "FNumber": 4.5,
        "Flash": 16,
        "FocalLength": 26,
        "FocalLengthIn35mmFormat": 39,
        "ISO": 3200,
        "InteropOffset": 36748,
        "LensModel": "E PZ 16-50mm F3.5-5.6 OSS",
        "LightSource": 10,
        "MaxApertureValue": 4.3359375,
        "MeteringMode": 5,
        "RecommendedExposureIndex": 3200,
        "Saturation": 0,
        "SceneCaptureType": 0,
        "SensitivityType": 2,
        "Sharpness": 0,
        "WhiteBalance": 1,
      },
      "gps": Object {},
      "image": Object {
        "ExifOffset": 360,
        "ImageDescription": "                               ",
        "Make": "SONY",
        "Model": "NEX-3N",
        "ModifyDate": "2014:02:05 18:06:22",
        "Orientation": 1,
        "ResolutionUnit": 2,
        "Software": "NEX-3N v1.00",
        "XResolution": 350,
        "YCbCrPositioning": 2,
        "YResolution": 350,
      },
    },
    "file": Object {
      "createdAt": 1543178776000,
      "modifiedAt": 1543178776000,
      "size": 5767168,
    },
  },
  "preview": Object {
    "fullSizePath": "fullSize/6c/4e0614f36b68ddf1ef24b938691e48723f71dc.jpg",
    "thumbnailPath": "thumbnail/6c/4e0614f36b68ddf1ef24b938691e48723f71dc.jpg",
  },
  "relativePath": "sonyNex3.jpg",
}
`;
